cornucopia/src/codegen.rs

```diff
@@ -278,36 +245,34 @@ fn composite_fromsql(
     name: &str,
     schema: &str,
 ) {
-    let field_names = join_comma(fields, |w, f| gen!(w, "{}", f.name));
-    let read_fields = join_ln(fields.iter().enumerate(), |w, (index, f)| {
-        gen!(
-            w,
-            "let _oid = postgres_types::private::read_be_i32(&mut out)?;
-            let {} = postgres_types::private::read_value(fields[{index}].type_(), &mut out)?;",
-            f.name,
-        );
-    });
-
-    gen!(
-        w,
-        r#"impl<'a> postgres_types::FromSql<'a> for {struct_name}Borrowed<'a> {{
-            fn from_sql(ty: &postgres_types::Type, out: &'a [u8]) -> 
-                Result<{struct_name}Borrowed<'a>, Box<dyn std::error::Error + Sync + Send>> 
-            {{
-                let fields = match *ty.kind() {{
+    let field_names = fields.iter().map(|p| &p.name);
+    let read_idx = 0..fields.len();
+    code!(w =>
+        impl<'a> postgres_types::FromSql<'a> for ${struct_name}Borrowed<'a> {
+            fn from_sql(ty: &postgres_types::Type, out: &'a [u8]) ->
+                Result<${struct_name}Borrowed<'a>, Box<dyn std::error::Error + Sync + Send>>
+            {
+                let fields = match *ty.kind() {
                     postgres_types::Kind::Composite(ref fields) => fields,
                     _ => unreachable!(),
-                }};
+                };
                 let mut out = out;
                 let num_fields = postgres_types::private::read_be_i32(&mut out)?;
-                {read_fields}
-                Ok({struct_name}Borrowed {{ {field_names} }})
-            }}
-
-            fn accepts(ty: &postgres_types::Type) -> bool {{
-                ty.name() == "{name}" && ty.schema() == "{schema}"
-            }}
-        }}"#
+                if num_fields as usize != fields.len() {
+                    return std::result::Result::Err(
+                        std::convert::Into::into(format!("invalid field count: {} vs {}", num_fields, fields.len())));
+                }
```

Comment:
Is this really needed? If the number of fields is wrong, its that our codegen implementation is bugged, and thus a runtime panic is the sensible thing to do (i.e. no need to pay a runtime cost for this).

Maybe I'm wrong though, please let me know :)

Reply:
I just follow the rust postgres macro https://github.com/sfackler/rust-postgres/blob/1e944c4ffc9e7674842fb86a8340187c731efdae/postgres-derive/src/fromsql.rs#L179.
We could argue that the generated code will always match the database schema, but I don't think these tests have a real impact on performance and may still be useful when the user changes the schema.

Reply:
We could benchmark an improved codegen without runtime test but I propose to do this in another PR

Reply:
> may still be useful when the user changes the schema.

That's a good point. hitting this may mean that our codegen is bugged (unlikely since we test our codegen) OR it might mean that the user did not regenerate their code after changing their schema.

Perhaps we could add a more explicit error message then (something along the lines of "did you forget to regenerate your query files?"). Or perhaps we could leave it as is to conform with rust postgres.

Reply:
If we change the error message here, we should change others too. I propose to leave it as is to conform with rust postgres for now and explore alternatives in another PR.